File: MsgServerClientSdk/SdkMsgClientAndroid/sdkmsgclient/src/main/java/org/dync/teameeting/sdkmsgclient/msgs/MsMsgUtil.java
package org.dync.teameeting.sdkmsgclient.msgs;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Created by hp on 7/13/16.
 */
public class MsMsgUtil {

    private static MSMessage JSONToMessage(JSONObject jobj) throws JSONException {
        MSMessage mMsg = new MSMessage();
        if (jobj.has("content"))
            mMsg.setContent(jobj.getString("content"));
        else
            mMsg.setContent("");
        if (jobj.has("extra"))
            mMsg.setExtra(jobj.getString("extra"));
        else
            mMsg.setExtra("");
        if (jobj.has("msgId"))
            mMsg.setMsgId(jobj.getString("msgId"));
        else
            mMsg.setMsgId("");
        if (jobj.has("toId"))
            mMsg.setToId(jobj.getString("toId"));
        else
            mMsg.setToId("");
        if (jobj.has("fromId"))
            mMsg.setFromId(jobj.getString("fromId"));
        else
            mMsg.setFromId("");
        if (jobj.has("groupId"))
            mMsg.setGroupId(jobj.getString("groupId"));
        else
            mMsg.setGroupId("");
        if (jobj.has("nickName"))
            mMsg.setNickName(jobj.getString("nickName"));
        else
            mMsg.setNickName("");
        if (jobj.has("uiconUrl"))
            mMsg.setUiconUrl(jobj.getString("uiconUrl"));
        else
            mMsg.setUiconUrl("");
        if (jobj.has("cash"))
            mMsg.setCash(jobj.getString("cash"));
        else
            mMsg.setCash("");
        if (jobj.has("wishcont"))
            mMsg.setWishcont(jobj.getString("wishcont"));
        else
            mMsg.setWishcont("");
        if (jobj.has("toNickName"))
            mMsg.setToNickName(jobj.getString("toNickName"));
        else
            mMsg.setToNickName("");
        if (jobj.has("millSec"))
            mMsg.setMillSec(jobj.getInt("millSec"));
        else
            mMsg.setMillSec(0);
        if (jobj.has("msgType"))
            mMsg.setMsgType(jobj.getInt("msgType"));
        else
            mMsg.setMsgType(0);
        if (jobj.has("flag"))
            mMsg.setFlag(jobj.getInt("flag"));
        else
            mMsg.setFlag(0);
        if (jobj.has("push"))
            mMsg.setPush(jobj.getInt("push"));
        else
            mMsg.setPush(0);
        return mMsg;
    }

    private static JSONObject MessageToJSON(MSMessage mobj) throws JSONException {
        JSONObject jsonObj = new JSONObject();
        jsonObj.put("content", mobj.getContent());
        jsonObj.put("extra", mobj.getExtra());
        jsonObj.put("msgId", mobj.getMsgId());
        jsonObj.put("toId", mobj.getToId());
        jsonObj.put("fromId", mobj.getFromId());
        jsonObj.put("groupId", mobj.getGroupId());
        jsonObj.put("nickName", mobj.getNickName());
        jsonObj.put("uiconUrl", mobj.getUiconUrl());
        jsonObj.put("cash", mobj.getCash());
        jsonObj.put("wishcont", mobj.getWishcont());
        jsonObj.put("toNickName", mobj.getToNickName());
        jsonObj.put("millSec", mobj.getMillSec());
        jsonObj.put("msgType", mobj.getMsgType());
        jsonObj.put("flag", mobj.getFlag());
        jsonObj.put("push", mobj.getPush());
        return jsonObj;
    }

/////////////////////////////////////////////////

    public static MSMessage Encode2JsonWithTxtMsg(MSSubMessage.MSTxtMessage txtMsg, int msgType) {
        MSMessage mMsg = new MSMessage();
        mMsg.setContent(txtMsg.getContent());
        mMsg.setGroupId(txtMsg.getGroupId());
        mMsg.setPush(txtMsg.getPush());
        mMsg.setToId(txtMsg.getToId());
        mMsg.setNickName(MsgClient.getInstance().getmStrNname());
        mMsg.setUiconUrl(MsgClient.getInstance().getmStrUicon());
        mMsg.setFromId(MsgClient.getInstance().getmStrUserId());
        mMsg.setToNickName(txtMsg.getToNickName());
        mMsg.setMsgType(msgType);

        return mMsg;
    }

    public static MSMessage Encode2JsonWithLivMsg(MSSubMessage.MSLivMessage livMsg, int msgType) {
        MSMessage mMsg = new MSMessage();
        mMsg.setGroupId(livMsg.getGroupId());
        mMsg.setToId(livMsg.getToId());
        mMsg.setPush(livMsg.getPush());
        mMsg.setNickName(MsgClient.getInstance().getmStrNname());
        mMsg.setUiconUrl(MsgClient.getInstance().getmStrUicon());
        mMsg.setFromId(MsgClient.getInstance().getmStrUserId());
        mMsg.setToNickName(livMsg.getToNickName());
        mMsg.setFlag(livMsg.getFlag());
        mMsg.setMsgType(msgType);

        return mMsg;
    }

    public static MSMessage Encode2JsonWithRenMsg(MSSubMessage.MSRenMessage renMsg, int msgType) {
        MSMessage mMsg = new MSMessage();
        mMsg.setGroupId(renMsg.getGroupId());
        mMsg.setToId(renMsg.getToId());
        mMsg.setPush(renMsg.getPush());
        mMsg.setNickName(MsgClient.getInstance().getmStrNname());
        mMsg.setUiconUrl(MsgClient.getInstance().getmStrUicon());
        mMsg.setFromId(MsgClient.getInstance().getmStrUserId());
        mMsg.setCash(renMsg.getCash());
        mMsg.setWishcont(renMsg.getWishcont());
        mMsg.setToNickName(renMsg.getToNickName());
        mMsg.setMsgType(msgType);

        return mMsg;
    }

    public static MSMessage Encode2JsonWithBlkMsg(MSSubMessage.MSBlkMessage blkMsg, int msgType) {
        MSMessage mMsg = new MSMessage();
        mMsg.setGroupId(blkMsg.getGroupId());
        mMsg.setToId(blkMsg.getToId());
        mMsg.setPush(blkMsg.getPush());
        mMsg.setNickName(MsgClient.getInstance().getmStrNname());
        mMsg.setUiconUrl(MsgClient.getInstance().getmStrUicon());
        mMsg.setFromId(MsgClient.getInstance().getmStrUserId());
        mMsg.setToNickName(blkMsg.getToNickName());
        mMsg.setFlag(blkMsg.getFlag());
        mMsg.setMsgType(msgType);

        return mMsg;
    }

    public static MSMessage Encode2JsonWithFbdMsg(MSSubMessage.MSFbdMessage fbdMsg, int msgType) {
        MSMessage mMsg = new MSMessage();
        mMsg.setGroupId(fbdMsg.getGroupId());
        mMsg.setToId(fbdMsg.getToId());
        mMsg.setPush(fbdMsg.getPush());
        mMsg.setNickName(MsgClient.getInstance().getmStrNname());
        mMsg.setUiconUrl(MsgClient.getInstance().getmStrUicon());
        mMsg.setFromId(MsgClient.getInstance().getmStrUserId());
        mMsg.setToNickName(fbdMsg.getToNickName());
        mMsg.setFlag(fbdMsg.getFlag());
        mMsg.setMsgType(msgType);

        return mMsg;
    }

    public static MSMessage Encode2JsonWithMgrMsg(MSSubMessage.MSMgrMessage mgrMsg, int msgType) {
        MSMessage mMsg = new MSMessage();
        mMsg.setGroupId(mgrMsg.getGroupId());
        mMsg.setToId(mgrMsg.getToId());
        mMsg.setPush(mgrMsg.getPush());
        mMsg.setNickName(MsgClient.getInstance().getmStrNname());
        mMsg.setUiconUrl(MsgClient.getInstance().getmStrUicon());
        mMsg.setFromId(MsgClient.getInstance().getmStrUserId());
        mMsg.setToNickName(mgrMsg.getToNickName());
        mMsg.setFlag(mgrMsg.getFlag());
        mMsg.setMsgType(msgType);

        return mMsg;
    }

    public static String JsonToString(MSMessage msg) {
        JSONObject jobj = null;
        try {
            jobj = MsMsgUtil.MessageToJSON(msg);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        if (null != jobj) {
            return jobj.toString();
        } else {
            return "";
        }
    }

    public static MSMessage DecodeJsonToMessage(String jsonStr) {

        JSONObject obj = null;
        try {
            obj = new JSONObject(jsonStr);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        if (null == obj) {
            System.err.println("MsMsgUtil DecodeJsonToMessage error, obj is null");
            return null;
        }
        MSMessage mMsg = null;
        try {
            mMsg = MsMsgUtil.JSONToMessage(obj);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return mMsg;
    }
}

File: MsgServer/MsgServerClient/ClientCommon/MSMsgUtil.cpp
//
//  MSMsgUtil.m
//  SdkMsgClient::os
//
//  Created by hp on 7/12/16.
//  Copyright © 2016 Dync. All rights reserved.
//

#include "MSMsgUtil.h"
#include "MsgClient.h"
#include "MSMessage.h"
#include "MSSubMessage.h"

#include "rapidjson/document.h"
#include "rapidjson/prettywriter.h"
#include "rapidjson/stringbuffer.h"


MSMessage*  MSMsgUtil::EncodeMessageWithTxtMsg(MSTxtMessage* txtMsg, int msgType)
{
    MSMessage* mMsg = new MSMessage();
    mMsg->SetContent(txtMsg->GetContent());
    mMsg->SetGroupId(txtMsg->GetGroupId());
    mMsg->SetPush(txtMsg->GetPush());
    mMsg->SetToId(txtMsg->GetToId());
    mMsg->SetNickName(MsgClient::Instance().MCGetNickName());
    mMsg->SetUiconUrl(MsgClient::Instance().MCGetUIconUrl());
    mMsg->SetFromId(MsgClient::Instance().MCGetUserId());
    mMsg->SetToNickName(txtMsg->GetToNickName());
    mMsg->SetMsgType((MCMsgType)msgType);

    return mMsg;
}

MSMessage*  MSMsgUtil::EncodeMessageWithLivMsg(MSLivMessage* livMsg, int msgType)
{
    MSMessage* mMsg = new MSMessage();
    mMsg->SetGroupId(livMsg->GetGroupId());
    mMsg->SetToId(livMsg->GetToId());
    mMsg->SetPush(livMsg->GetPush());
    mMsg->SetNickName(MsgClient::Instance().MCGetNickName());
    mMsg->SetUiconUrl(MsgClient::Instance().MCGetUIconUrl());
    mMsg->SetFromId(MsgClient::Instance().MCGetUserId());
    mMsg->SetToNickName(livMsg->GetToNickName());
    mMsg->SetFlag(livMsg->GetFlag());
    mMsg->SetMsgType((MCMsgType)msgType);

    return mMsg;
}

MSMessage*  MSMsgUtil::EncodeMessageWithRenMsg(MSRenMessage* renMsg, int msgType)
{
    MSMessage* mMsg = new MSMessage();
    mMsg->SetGroupId(renMsg->GetGroupId());
    mMsg->SetToId(renMsg->GetToId());
    mMsg->SetPush(renMsg->GetPush());
    mMsg->SetNickName(MsgClient::Instance().MCGetNickName());
    mMsg->SetUiconUrl(MsgClient::Instance().MCGetUIconUrl());
    mMsg->SetFromId(MsgClient::Instance().MCGetUserId());
    mMsg->SetCash(renMsg->GetCash());
    mMsg->SetWishcont(renMsg->GetWishcont());
    mMsg->SetToNickName(renMsg->GetToNickName());
    mMsg->SetMsgType((MCMsgType)msgType);

    return mMsg;
}

MSMessage*  MSMsgUtil::EncodeMessageWithBlkMsg(MSBlkMessage* blkMsg, int msgType)
{
    MSMessage* mMsg = new MSMessage();
    mMsg->SetGroupId(blkMsg->GetGroupId());
    mMsg->SetToId(blkMsg->GetToId());
    mMsg->SetPush(blkMsg->GetPush());
    mMsg->SetNickName(MsgClient::Instance().MCGetNickName());
    mMsg->SetUiconUrl(MsgClient::Instance().MCGetUIconUrl());
    mMsg->SetFromId(MsgClient::Instance().MCGetUserId());
    mMsg->SetToNickName(blkMsg->GetToNickName());
    mMsg->SetFlag(blkMsg->GetFlag());
    mMsg->SetMsgType((MCMsgType)msgType);

    return mMsg;
}

MSMessage*  MSMsgUtil::EncodeMessageWithFbdMsg(MSFbdMessage* fbdMsg, int msgType)
{
    MSMessage* mMsg = new MSMessage();
    mMsg->SetGroupId(fbdMsg->GetGroupId());
    mMsg->SetToId(fbdMsg->GetToId());
    mMsg->SetPush(fbdMsg->GetPush());
    mMsg->SetNickName(MsgClient::Instance().MCGetNickName());
    mMsg->SetUiconUrl(MsgClient::Instance().MCGetUIconUrl());
    mMsg->SetFromId(MsgClient::Instance().MCGetUserId());
    mMsg->SetToNickName(fbdMsg->GetToNickName());
    mMsg->SetFlag(fbdMsg->GetFlag());
    mMsg->SetMsgType((MCMsgType)msgType);

    return mMsg;
}

MSMessage*  MSMsgUtil::EncodeMessageWithMgrMsg(MSMgrMessage* mgrMsg, int msgType)
{
    MSMessage* mMsg = new MSMessage();
    mMsg->SetGroupId(mgrMsg->GetGroupId());
    mMsg->SetToId(mgrMsg->GetToId());
    mMsg->SetPush(mgrMsg->GetPush());
    mMsg->SetNickName(MsgClient::Instance().MCGetNickName());
    mMsg->SetUiconUrl(MsgClient::Instance().MCGetUIconUrl());
    mMsg->SetFromId(MsgClient::Instance().MCGetUserId());
    mMsg->SetToNickName(mgrMsg->GetToNickName());
    mMsg->SetFlag(mgrMsg->GetFlag());
    mMsg->SetMsgType((MCMsgType)msgType);

    return mMsg;
}

MSMessage*  MSMsgUtil::JsonToMessage(const std::string& jsonMsg)
{
    if (jsonMsg.length()==0) return nullptr;
    std::string     content("");
    std::string     extra("");
    std::string     msgId("");
    std::string     toId("");
    std::string     fromId("");
    std::string     groupId("");
    std::string     nickName("");
    std::string     uiconUrl("");
    std::string     cash("");
    std::string     wishcont("");
    std::string     toNickName("");
    unsigned int    millSec=0;
    MCMsgType       msgType=MCMsgTypeTtxt;
    int             flag=0;
    int             push=0;

    rapidjson::Document jsonMsgDoc;
    if (!jsonMsgDoc.ParseInsitu<0>((char*)jsonMsg.c_str()).HasParseError())
    {
        if (jsonMsgDoc.HasMember("content")&&jsonMsgDoc["content"].IsString())
        {
            content = jsonMsgDoc["content"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("extra")&&jsonMsgDoc["extra"].IsString())
        {
            extra = jsonMsgDoc["extra"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("msgId")&&jsonMsgDoc["msgId"].IsString())
        {
            msgId = jsonMsgDoc["msgId"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("toId")&&jsonMsgDoc["toId"].IsString())
        {
            toId = jsonMsgDoc["toId"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("fromId")&&jsonMsgDoc["fromId"].IsString())
        {
            fromId = jsonMsgDoc["fromId"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("groupId")&&jsonMsgDoc["groupId"].IsString())
        {
            groupId = jsonMsgDoc["groupId"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("nickName")&&jsonMsgDoc["nickName"].IsString())
        {
            nickName = jsonMsgDoc["nickName"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("uiconUrl")&&jsonMsgDoc["uiconUrl"].IsString())
        {
            uiconUrl = jsonMsgDoc["uiconUrl"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("cash")&&jsonMsgDoc["cash"].IsString())
        {
            cash = jsonMsgDoc["cash"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("wishcont")&&jsonMsgDoc["wishcont"].IsString())
        {
            wishcont = jsonMsgDoc["wishcont"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("toNickName")&&jsonMsgDoc["toNickName"].IsString())
        {
            toNickName = jsonMsgDoc["toNickName"].GetString();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("millSec")&&jsonMsgDoc["millSec"].IsInt())
        {
            millSec = jsonMsgDoc["millSec"].GetInt();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("msgType")&&jsonMsgDoc["msgType"].IsInt())
        {
            msgType = (MCMsgType)jsonMsgDoc["msgType"].GetInt();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("flag")&&jsonMsgDoc["flag"].IsInt())
        {
            flag = jsonMsgDoc["flag"].GetInt();
        } else {
            return nullptr;
        }
        if (jsonMsgDoc.HasMember("push")&&jsonMsgDoc["push"].IsInt())
        {
            push = jsonMsgDoc["push"].GetInt();
        } else {
            return nullptr;
        }
    } else {
        return nullptr;
    }
    MSMessage *msg = new MSMessage();
    msg->SetContent(content);
    msg->SetExtra(extra);
    msg->SetMsgId(msgId);
    msg->SetToId(toId);
    msg->SetFromId(fromId);
    msg->SetGroupId(groupId);
    msg->SetNickName(nickName);
    msg->SetUiconUrl(uiconUrl);
    msg->SetCash(cash);
    msg->SetWishcont(wishcont);
    msg->SetToNickName(toNickName);
    msg->SetMillSec(millSec);
    msg->SetMsgType((MCMsgType)msgType);
    msg->SetFlag(flag);
    msg->SetPush(push);
    return msg;
}

std::string*    MSMsgUtil::MessageToJson(MSMessage& msg)
{
    rapidjson::Document jDoc;
    rapidjson::StringBuffer sb;
    rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
    jDoc.SetObject();
    jDoc.AddMember("content", msg.GetContent().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("extra", msg.GetExtra().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("msgId", msg.GetMsgId().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("toId", msg.GetToId().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("fromId", msg.GetFromId().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("groupId", msg.GetGroupId().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("nickName", msg.GetNickName().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("uiconUrl", msg.GetUiconUrl().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("cash", msg.GetCash().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("wishcont", msg.GetWishcont().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("toNickName", msg.GetToNickName().c_str(), jDoc.GetAllocator());
    jDoc.AddMember("millSec", msg.GetMillSec(), jDoc.GetAllocator());
    jDoc.AddMember("msgType", msg.GetMsgType(), jDoc.GetAllocator());
    jDoc.AddMember("flag", msg.GetFlag(), jDoc.GetAllocator());
    jDoc.AddMember("push", msg.GetPush(), jDoc.GetAllocator());

    jDoc.Accept(writer);
    return new std::string(sb.GetString());
}
